pydantic_ai_graph/pydantic_ai_graph/state.py

```diff
@@ -19,33 +24,40 @@ def serialize(self) -> bytes | None:
         raise NotImplementedError
 
 
+RunEndT = TypeVar('RunEndT', default=None)
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
 StateT = TypeVar('StateT', bound=Union[None, AbstractState], default=None)
 
 
 @dataclass
-class Snapshot:
-    """Snapshot of a graph."""
-
-    last_node_id: str
-    next_node_id: str
-    start_ts: datetime
-    duration: float
-    state: bytes | None = None
```

Comment:
I replaced this with non-serialized history events mostly because I think people will want to interact with the deserialized history, so storing it as raw bytes seems unfortunate. I think we need to ensure both `Step` and `EndEvent` are serializable (and therefore all `BaseNode` instances and all `RunEndT` values need to be serializable). I'm not sure the best way to achieve that though.